src/execution/operator/scan/physical_table_scan.cpp

```diff
@@ -101,9 +101,32 @@ SourceResultType PhysicalTableScan::GetData(ExecutionContext &context, DataChunk
 
 	TableFunctionInput data(bind_data.get(), l_state.local_state.get(), g_state.global_state.get());
 
-	if (function.function) {
-		function.function(context.client, data, chunk);
-		return chunk.size() == 0 ? SourceResultType::FINISHED : SourceResultType::HAVE_MORE_OUTPUT;
+	if (function.HasSimpleScan()) {
+		auto res = function.SimpleScan(context.client, data, chunk);
+
+		if (res == SourceResultType::BLOCKED) {
+			D_ASSERT(data.async_result.HasTasks());
+			auto guard = g_state.Lock();
+			auto inner_result = g_state.BlockSource(guard, input.interrupt_state);
+			if (inner_result == SourceResultType::FINISHED) {
+				return SourceResultType::FINISHED;
+			} else if (inner_result == SourceResultType::BLOCKED) {
+				AsyncResult async_handler(std::move(data.async_result));
+				async_handler.ScheduleTasks(input.interrupt_state, context.pipeline->executor);
```

Comment:
I'm still not convinced we can use `g_state.BlockSource` here, if we're then also handing the interrupt state to this `ScheduleTasks` method.

If the scheduled tasks are responsible for unblocking the task, we only need `g_state` to check whether we're allowed to block (i.e: query isn't cancelled), but it shouldn't take ownership of the interrupt state, which is what `BlockSource` does do.